### ISSUE TITLE
Introduce "delegatesFocus" flag to shadow root

### ISSUE THREAD

**Issue by TakayoshiKochi** [WICG/webcomponents#126]
To fix focusability and focus navigation order issue ([W3C 25473](https://www.w3.org/Bugs/Public/show_bug.cgi?id=25473), [Chromium 380445](https://code.google.com/p/chromium/issues/detail?id=380445)), introduce `delegatesFocus` flag under shadow root, to address some shortcomings that
the default focus navigation behavior ([defined in the Shadow DOM spec](https://w3c.github.io/webcomponents/spec/shadow/#focus-navigation)) has.

Detailed doc is here:
https://github.com/TakayoshiKochi/tabindex-focus-navigation-explainer/blob/master/TabindexFocusNavigationExplainer.md

I'll post the proposal doc soon.
This is experimentally implmented in Blink [Chromium 496005](http://crbug.com/496005)

**Comment by rniwa** [comment]
Why can't we always use `autofocus` attribute to infer it?  That is, behave as if `delegatesFocus` as `true` if and only if the shadow DOM contains an element with `autofocus` attribute but don't expose `delegatesFocus`.

**Comment by rniwa** [comment]
@annevk

**Comment by annevk** [comment]
I don't see how that solves the problem of skipping the host element adequately.

**Comment by rniwa** [comment]
I'm saying that skip the host element if its shadow DOM contains an element with `autofocus` attribute.

**Comment by annevk** [comment]
But if it only contains a single `<input>` you would not skip it? Why would a component always want to use `autofocus`?

**Comment by TakayoshiKochi** [comment]
Let me summarize the status of this issue to get the ball rolling again.
1. @rniwa's question about `autofocus` (#126) - as @annevk already replied, why `autofocus` would suffice for what `delegatesFocus` tries to solve is unclear - @rniwa, could you add more comment on this?
2. (#105) - about auto focus-sliding into shadow tree behavior. I think my use case is understood by @annevk but @rniwa's comment that `delegatesFocus` is unnecessary isn't well understood by me or @annevk.  @rniwa, could you add more comment?
3. (#276) - focus processing model is still unclear - I will have to come up with more clear model of focus processing as a patch to current HTML spec, am I understanding your concern right? @annevk

**Comment by annevk** [comment]
Yes, I would like to see how shadow DOM affects the focus model we have today (and even that likely has holes still). After that is explained it would be easier to reason about any changes to it.

**Comment by hayatoito** [comment]
@TakayoshiKochi , could you summarize the status of this issue? This is a blocking issue for v1.

**Comment by rniwa** [comment]
Never mind about my proposal.  I was misunderstanding the problem at the time.

Now that I understand the problem better, I think it's fine to have a flag.  If we're going to have `setForceFocusable` though, we should probably call this flag `focusable`.  Flipping the boolean in the init dictionary and the method argument is obnoxious.

**Comment by rniwa** [comment]
Also, do we really need `setForceFocusable`?  What are use cases in which whether the shadow root should/can be focused change dynamically?

**Comment by TakayoshiKochi** [comment]
The `focusable` or `setForceFocusable` were introduced to the document trying to explain some of the
native elements via shadow DOM.

For example `<a name=...>` isn't focusable but `<a href=...>` is focusable, without having to add/remove `tabindex` attribute in HTML.  So it could be used for emulating such behavior, but I don't think it is a viable use case, and we can omit it at least for v1.

**Comment by TakayoshiKochi** [comment]
Thus, the changes I would like to introduce is basically listed [here](https://github.com/TakayoshiKochi/tabindex-focus-navigation-explainer/blob/master/TabindexFocusNavigationExplainer.md#delegatesfocus-details),
by giving `delegatesFocus: true` in `attachShadow()` will affect:
1. if shadow host has 0 or positive `tabindex` attribute value, the sequential navigation order honors that value, but actual focus goes into an element in its shadow (depending on navigation order, first or last focusable element in shadow will be picked)
2. `focus()` on a shadow host will also delegate focus to its shadow
3. `autofocus` attribute on a shadow host will also delegate focus to its shadow
4. If non-focusable area in shadow is clicked, shadow host gets focus but then delegates focus to its shadow
5. CSS `:focus` pseudo class on shadow host should match when its shadow or its descendant has focus

@rniwa @travisleithead @annevk any opinion for each item above?

**Comment by annevk** [comment]
I think 5 should also be recursive, no?

Will we do something similar for slots?

**Comment by TakayoshiKochi** [comment]
Re `:focus` matching, it should be recursive.  e.g. in the following example,
`div#host1:focus` should match when either `input#outer` or `input#inner` is focused.
Of course `div#host2:focus` only matches when `input#inner` is focused.

``` html
<div#host1>
  <:shadow>
     <div#host2>
        <:shadow>
           <input#inner>
        </:shadow>
    <input#outer>
  </:shadow>
</div>
```

Because a slot is not focusable, I don't think identical mechanism is necessary, but regarding to `:focus` matching I think it depends on the result of how we treat focus in slot for `.activeElement` (#358).

**Comment by rniwa** [comment]
Let's not block this issue on the issue #358 or the issue #308.  I think the proposal in https://github.com/w3c/webcomponents/issues/126#issuecomment-188160444 is sound.

**Comment by TakayoshiKochi** [comment]
I start working on updating the specs.
